--- a/tools/hciattach.h	2021-03-23 01:57:47.389906065 +0000
+++ b/tools/hciattach.h	2021-03-23 02:00:14.055175169 +0000
@@ -67,5 +67,5 @@ int ath3k_init(int fd, int speed, int in
 int ath3k_post(int fd, int pm);
 int qualcomm_init(int fd, int speed, struct termios *ti, const char *bdaddr);
 int intel_init(int fd, int init_speed, int *speed, struct termios *ti);
-int bcm43xx_init(int fd, int def_speed, int speed, struct termios *ti,
+int bcm43xx_init(int fd, int proto, int def_speed, int speed, struct termios *ti,
 		const char *bdaddr);
--- a/tools/hciattach_bcm43xx.c	2021-03-23 01:28:14.758632417 +0000
+++ b/tools/hciattach_bcm43xx.c	2021-03-23 01:40:47.849080263 +0000
@@ -44,7 +44,7 @@
 #include "hciattach.h"
 
 #ifndef FIRMWARE_DIR
-#define FIRMWARE_DIR "/etc/firmware"
+#define FIRMWARE_DIR "/lib/firmware"
 #endif
 
 #define FW_EXT ".hcd"
@@ -352,7 +352,7 @@ static int bcm43xx_locate_patch(const ch
 	return ret;
 }
 
-int bcm43xx_init(int fd, int def_speed, int speed, struct termios *ti,
+int bcm43xx_init(int fd, int proto, int def_speed, int speed, struct termios *ti,
 		const char *bdaddr)
 {
 	char chip_name[20];
@@ -369,8 +369,10 @@ int bcm43xx_init(int fd, int def_speed,
 	if (bcm43xx_locate_patch(FIRMWARE_DIR, chip_name, fw_path)) {
 		fprintf(stderr, "Patch not found, continue anyway\n");
 	} else {
-		if (bcm43xx_set_speed(fd, ti, speed))
-			return -1;
+		/* If uart proto != HCI_UART_3WIRE, only then set device speed */ 
+		if( proto != HCI_UART_3WIRE )
+			if (bcm43xx_set_speed(fd, ti, speed))
+				return -1;
 
 		if (bcm43xx_load_firmware(fd, fw_path))
 			return -1;
--- a/tools/hciattach.c	2021-03-23 01:24:32.016752748 +0000
+++ b/tools/hciattach.c	2021-03-23 01:27:55.290467557 +0000
@@ -276,7 +276,7 @@ static int intel(int fd, struct uart_t *
 
 static int bcm43xx(int fd, struct uart_t *u, struct termios *ti)
 {
-	return bcm43xx_init(fd, u->init_speed, u->speed, ti, u->bdaddr);
+	return bcm43xx_init(fd, u->proto, u->init_speed, u->speed, ti, u->bdaddr);
 }
 
 static int read_check(int fd, void *buf, int count)
@@ -1091,6 +1091,9 @@ struct uart_t uart[] = {
 	{ "bcm43xx",    0x0000, 0x0000, HCI_UART_H4,   115200, 3000000,
 				FLOW_CTL, DISABLE_PM, NULL, bcm43xx, NULL  },
 
+	{ "bcm43xx-3wire",    0x0000, 0x0000, HCI_UART_3WIRE,   115200, 3000000,
+				0, DISABLE_PM, NULL, bcm43xx, NULL  },
+
 	{ "ath3k",    0x0000, 0x0000, HCI_UART_ATH3K, 115200, 115200,
 			FLOW_CTL, DISABLE_PM, NULL, ath3k_ps, ath3k_pm  },
 
--- a/tools/bdaddr.c	2021-03-23 02:12:28.285506681 +0000
+++ b/tools/bdaddr.c	2021-03-23 02:15:01.006818664 +0000
@@ -316,6 +316,7 @@ static struct {
 	{ 48,		st_write_bd_addr,	generic_reset_device	},
 	{ 57,		ericsson_write_bd_addr,	generic_reset_device	},
 	{ 72,		mrvl_write_bd_addr,	generic_reset_device	},
+	{ 305,		bcm_write_bd_addr,	generic_reset_device	},
 	{ 65535,	NULL,			NULL			},
 };
 
